### PR TITLE
Add Gateway Protocol

### PR DESCRIPTION
**Pull request by chris124567**
Added module similar to `net/rhp` that instead uses the gateway handshake and `net/mux`.  Refactored some of the similarities between rhp and the gateway into `net/rpc`.

### CODE REVIEW

File: net/rhp/session_test.go [hunk 1]
```diff
@@ -14,10 +14,11 @@ import (
 
 	"lukechampine.com/frand"
 
+	"go.sia.tech/core/net/rpc"
```
Comment (lukechampine) [inline net/rhp/session_test.go:17]: looks like this package wasn't added? I don't understand why this didn't cause a CI failure 🤔

File: net/gateway/gateway.go [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+		}
+	}
+}
+
+func (s *Session) writeMessage(obj rpc.ProtocolObject) error {
+	if err := s.PrematureCloseErr(); err != nil {
+		return err
+	}
+
+	stream, err := s.mux.DialStream()
```
Comment (lukechampine) [inline net/gateway/gateway.go:66]: we want a 1-to-1 mapping between mux streams and RPCs; the flow for an RPC is:
1. create a new mux stream
2. send rpc ID
3. send rpc message body
4. read rpc response
5. close mux stream

So I think the right approach is to move the `DialStream` into `WriteMessage`, and have `writeMessage` operate on the stream.

File: net/gateway/gateway.go [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+	if err := s.writeMessage(&rpc.RPCResponse{re, resp}); err != nil {
+		return fmt.Errorf("WriteResponse: %w", err)
+	}
+	return nil
+}
+
+// ReadResponse reads an RPC response. If the response is an error, it is
+// returned directly.
+func (s *Session) ReadResponse(resp rpc.ProtocolObject, maxLen uint64) error {
+	rr := rpc.RPCResponse{nil, resp}
```
Comment (lukechampine) [inline net/gateway/gateway.go:144]: not sure how I feel about a common RPC package (I'll wait until I see it), but assuming we go that route, we'll want to rename `RPCError` -> `Error` and `RPCResponse` -> `Response`

Comment (lukechampine) [reply net/gateway/gateway.go:144]: hmm, or maybe put the `rpc` code in the top-level `net`, and keep the prefixes? 🤔 That feels more appropriate.

File: net/gateway/netaddress.go [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+)
+
+// MaxEncodedNetAddressLength is the maximum length of a NetAddress encoded
+// with the encode package. 266 was chosen because the maximum length for the
+// hostname is 254 + 1 for the separating colon + 5 for the port + 8 byte
+// string length prefix.
+const MaxEncodedNetAddressLength = 266
+
+// A NetAddress contains the information needed to contact a peer.
+type NetAddress string
```
Comment (lukechampine) [inline net/gateway/netaddress.go:17]: nah we're not reusing any of this code lol. Just use `string` for now, and we'll deal with the `IsLocal`/`IsLoopback`/etc. stuff later when we implement that actual gateway.

File: types/encoding.go [hunk 5]
```diff
@@ -66,6 +66,12 @@ func (e *Encoder) WritePrefix(i int) { e.WriteUint64(uint64(i)) }
 // WriteTime writes a time.Time value to the underlying stream.
 func (e *Encoder) WriteTime(t time.Time) { e.WriteUint64(uint64(t.Unix())) }
 
+// WriteString writes a string value to the underlying stream.
+func (e *Encoder) WriteString(s string) {
+	e.WritePrefix(len(s))
+	e.Write([]byte(s))
+}
```
Comment (lukechampine) [inline types/encoding.go:73]: 👍🏻

File: net/gateway/connect.go [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+	// listener / tcp connections not implemented yet
+	dummyAddr = "127.0.0.1:1"
+)
+
+type sessionHeader struct {
+	publicKey  ed25519.PublicKey
+	genesisID  types.BlockID
+	uniqueID   GatewayID
+	netAddress NetAddress
+}
```
[the remark this answers is not in the document]
Comment (lukechampine) [reply net/gateway/connect.go:28]: (In general, remember that we're writing "dumb" library code here; it should be making very few decisions internally. Let the caller worry about those decisions.)

File: net/gateway/connect.go [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+		genesisID:  types.GenesisBlock.ID(),
+		uniqueID:   g.uniqueID,
+		netAddress: g.netAddress,
+	}
+	if _, err := exchangeRemoteHeader(conn, ourHeader); err != nil {
+		return nil, err
+	}
+	if err := exchangeOurHeader(conn, ourHeader); err != nil {
+		return nil, err
+	}
```
Comment (lukechampine) [inline net/gateway/connect.go:99]: idk if it's necessary for these to all be separate functions like they were in the old codebase. Let's try inlining them all.

File: net/gateway/gateway.go (hunk 2 above)
Comment (n8maninger) [reply net/gateway/gateway.go:66]: It's an additional string handshake:
https://gitlab.com/NebulousLabs/siamux/-/blob/master/siamux.go#L301
https://github.com/SiaFoundation/siad/blob/master/modules/host/network.go#L250

https://github.com/lukechampine/us/pull/178

**Comment by chris124567** [comment]
Tried to only do one DialStream call in WriteRequest so we just reuse the stream but this strangely seems to make the subsequent ReadRequest call hang.  I'll keep trying.

Also

>IIRC we only needed to exchange keys so that we could encrypt the connection. mux handles that for us now, so we can ditch this field.

But don't we need their public key when we call dial?

File: net/gateway/gateway.go [hunk 8]
```diff
@@ -0,0 +1,157 @@
+package gateway
+
+import (
+	"fmt"
+	"go.sia.tech/core/net"
+	"go.sia.tech/core/net/mux"
+	"go.sia.tech/core/types"
+	"io"
+	gnet "net"
```
Comment (lukechampine) [inline net/gateway/gateway.go:9]: ehh yeah having to rename a stdlib package kinda sucks. Might be ok if it's rare to import both `net` packages, I guess? But this is definitely evidence against my suggestion.

File: net/gateway/connect.go [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+	acceptResponse       = "accept"
+	// listener / tcp connections not implemented yet
+	dummyAddr = "127.0.0.1:1"
+)
+
+type SessionConfig struct {
+	publicKey  ed25519.PublicKey
+	genesisID  types.BlockID
+	uniqueID   GatewayID
+	netAddress string
```
Comment (lukechampine) [inline net/gateway/connect.go:30]: these fields will need to be exported as well.
`GatewayID` can just be `[8]byte`, I think a separate type is overkill.

**Pull request merged**